packages/sitecore-jss/src/personalization/personalization-decisions-service.ts

```diff
@@ -0,0 +1,208 @@
+
+export type DataFetcherResolver = <T>(config: {
+  /**
+   * The request timeout in milliseconds
+   */
+  timeout?: number;
+}) => HttpDataFetcher<T>;
+
+export type RestPersonalizationDecisionsServiceConfig = {
+  /**
+   * Hostname of decisions service; e.g. http://my.site.core; Default: '', same host as a page
+   */
+  host?: string;
+  /**
+   * Relative path from host to decisions service. Default: /sitecore/api/layout/personalization/decision
+   */
+  route?: string;
+  /**
+   * This value overrides the default service URL.
+   * Note: `host` and `route` options are ignored if `serviceUrl` is set.
+   */
+  serviceUrl?: string;
+  /**
+   * The Sitecore SSC API key your app uses
+   */
+  apiKey: string;
+  /**
+   * The JSS application name
+   */
+  siteName: string;
+  /**
+   * Query string parameters of the current page to pass with request to decision service
+   */
+  currentPageParamsToTrack?: string[];
+  /**
+   * Enables/disables analytics tracking for the Layout Service invocation (default is true).
+   * More than likely, this would be set to false for SSG/hybrid implementations, and the
+   * JSS tracker would instead be used on the client-side: {@link https://jss.sitecore.com/docs/fundamentals/services/tracking}
+   * @default true
+   */
+  tracking?: boolean;
+  /**
+   * The request timeout in milliseconds
+   */
+  timeout?: number;
+  /**
+   * Data fetcher resolver in order to provide custom data fetcher
+   * @see DataFetcherResolver
+   * @see HttpDataFetcher<T>
+   * @see AxiosDataFetcher used by default
+   * @param {IncomingMessage} [req] Request instance
+   * @param {ServerResponse} [res] Response instance
+   */
+  dataFetcherResolver?: DataFetcherResolver;
+};
+
+/**
+ * note: encodeURIComponent is available via browser (window) or natively in node.js
+ * if you use another js engine for server-side rendering you may not have native encodeURIComponent
+ * and would then need to install a package for that functionality
+ * @param {Object} params
+ */
+function getQueryString(params: { [key: string]: unknown }) {
+  return Object.keys(params)
+    .map((k) => `${encodeURIComponent(k)}=${encodeURIComponent(params[k] as string)}`)
+    .join('&');
+}
+
+class ResponseError extends Error {
+  response: HttpResponse<unknown>;
+
+  constructor(message: string, response: HttpResponse<unknown>) {
+    super(message);
+
+    Object.setPrototypeOf(this, ResponseError.prototype);
+    this.response = response;
+  }
+}
+
+/**
+ * @param {HttpResponse<T>} response
+ * @throws {ResponseError} if response code is not ok
+ */
+function checkStatus<T>(response: HttpResponse<T>) {
+  if (response.status >= 200 && response.status < 300) {
+    return response;
+  }
+
+  const error = new ResponseError(response.statusText, response);
+  throw error;
+}
+
+/**
+ * @param {string} url
+ * @param {HttpDataFetcher} fetcher
+ * @param {Object} params
+ * @param {Object} data to POST with the request
+ */
+function fetchData<T>(
```

Comment:
Private file functions could be moved to the end of the file to simplify reading.

Reply:
removed